packages/teuchos/comm/test/Comm/teuchosSubcommTest.cpp

```diff
@@ -0,0 +1,42 @@
+// Test to exercise fix to Teuchos::Comm's createSubcommunicator, which
+// had leaked memory.
+// The fix added MPI_Comm_free to the opaqueWrappers. 
+// 8/2018  This test hangs of platform waterman.
+
+#include <stdio.h>
+#include <mpi.h>
+#include "Teuchos_CommHelpers.hpp"
+#include "Teuchos_DefaultComm.hpp"
+#include "Teuchos_RCP.hpp"
+#include "Teuchos_ArrayView.hpp"
+
+
+int main(int narg, char **arg)
+{
+  Teuchos::GlobalMPISession mpiSession(&narg,&arg);
+
+  Teuchos::RCP<const Teuchos::Comm<int> >
+    comm = Teuchos::DefaultComm<int>::getComm();
+  int me = comm->getRank();
+  int np = comm->getSize();
+
+  int niter = 4;
+  int *ids = NULL;
+  ids = new int[np/2+1];
+  ids[0] = me;
+  for (int i = 1; i < np/2+1; i++) {
+    ids[i] = (i != me ? i : 0);
+  }
+  Teuchos::ArrayView<const int> list(ids, np/2+1);
+  for (int i = 0; i < niter; i++) {
+    Teuchos::RCP<const Teuchos::Comm<int> > a 
+                                            = comm->createSubcommunicator(list);
```

Comment:
Both `createSubcommunicator()` and `duplicate()` create an RCP with a custom deleter.  The deleter checks whether `MPI_Finalize` has been called before attempting to call `MPI_Comm_free` on the RCP's internal `MPI_Comm`.  See [the implementation here](https://github.com/trilinos/Trilinos/blob/b783a65a6faed6a40372d404080ba2834f6af155/packages/teuchos/comm/src/Teuchos_DefaultMpiComm.hpp#L1747), and [the implementation of `safeCommFree` here](https://github.com/trilinos/Trilinos/blob/b783a65a6faed6a40372d404080ba2834f6af155/packages/teuchos/comm/src/Teuchos_DefaultMpiComm.cpp#L80).  The FIXME in the latter code should only apply if the _result_ of `createSubcommunicator()` or `duplicate()` outlives the `MPI_Finalize` call, but see my note below about the MPI Standard not explicitly forbidding failure to call `MPI_Comm_free`.